### ISSUE TITLE
I want user-contributed alt text to be available on filesets and in the Universal Viewer

### ISSUE THREAD

**Issue by nhomenda** [samvera/hyrax#7236]
### Descriptive summary

Make filesets and Universal Viewer display user-contributed alternative text to improve accessibility of the public view.

### Steps to reproduce the behavior in User Interface (UI)

1. Navigate to an image work that has an item/fileset with a thumbnail image.
2. Currently no alt text is available for the item/fileset's thumbnail on the work's show page or for the image in the UV

### Acceptance Criteria/Expected Behavior

- [ ] Alt text available in item/fileset thumbnails
- [ ] Alt text available in Universal Viewer

### Rationale (for feature request only)

Achieving WCAG 2.2 AA requires alt text for images beyond filename.

### Related work

Bill Cox's explanation of OSU's implementation with attached patch files:

I went back and dusted off the commits for the alt text work and attached a few files:

- [osul-alt-text.patch](https://github.com/user-attachments/files/22784696/osul-alt-text.patch) haș the changes we made locally as an app built on hyrax
- [hyrax-alt-text.patch](https://github.com/user-attachments/files/22784697/hyrax-alt-text.patch) has my untested attempt to translate that to hyrax itself
- [uv-alt-text.patch](https://github.com/user-attachments/files/22784695/uv-alt-text.patch) has just our local changes to UV

On the front-end, we updated UV and the image view.
In UV, when creating the viewer `<div>`, we add an `aria-label` attribute with the alt-text. From the commit history I see that we also tried adding `role=“img”`, but we removed that after testing with Windows Narrator. To support this, we added the alt text to the `IIIF::Presentation::ImageResource` when creating the manifest.
In Rails, we overrode `app/views/hyrax/file_sets/media_display/_image.html.erb` to use `file_set.alt_text` and `app/views/hyrax/file_sets/_form.html.erb` to allow setting the alt text.

The rest of the plumbing was adding the property to `::FileSet` and including it in `to_solr`, adding `::SolrDocument#alt_text` to expose the indexed field, and creating our own `::FileSetPresenter` to delegate the alt text to the solr document.

I saw online that there was some discussion about the predicate to use for the alt text. For our purposes, we decided to side-step that issue and created a dummy vocabulary:
```
VOCAB = ::RDF::Vocabulary.new("https://example.com/vocab/")
property :alt_text, predicate: VOCAB.alt_text, multiple: false
```
We also updated the configuration to include the alt text in the edit form and to use our presenter for filesets. The presenter configuration for filesets isn’t as centralized as it is for generic works, so we had to set it in multiple places and override `Hyrax::Forms::WorkForm` to allow configuring it.
```
Hyrax::Forms::FileSetEditForm.terms += %i[alt_text]

Hyrax::FileSetsController.show_presenter = ::FileSetPresenter
Hyrax::Forms::WorkForm.file_presenter_class = ::FileSetPresenter
Hyrax::MemberPresenterFactory.file_presenter_class = ::FileSetPresenter
Hyrax::PcdmMemberPresenterFactory.file_presenter_class = ::FileSetPresenter
Hyrax::SingleUseLinksViewerController.presenter_class = ::FileSetPresenter
```
We followed this pattern when added captions and transcripts.

The biggest issue is that this currently only sets the alt text once for the viewer. This handles a lot of our use cases but additional work is needed to handle multiple images in the same fileset.

**Comment by christhepianist** [comment]
[like]  Christine Peterson reacted to your message:
________________________________
From: Nick Homenda ***@***.***>
Sent: Wednesday, October 8, 2025 7:37:38 PM
To: samvera/hyrax ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [samvera/hyrax] I want user-contributed alt text to be available on filesets and in the Universal Viewer (Issue #7236)

[https://avatars.githubusercontent.com/u/20046899?s=20&v=4]nhomenda created an issue (samvera/hyrax#7236)<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_samvera_hyrax_issues_7236&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=UnL802G5SdVV-32R3u-ZQCPy_3ItQxrnxxLobtXEX90&m=F0sAgN-UWc1AJzrRPMBVHsNdIsFUCeCo9y_nJN4_DTwQE7oQjNFzpU7NsdjZWFpu&s=eV1TkqJoGeT1SaMgVEFrL5IfUi_boZV6QWiZs-bjSo8&e=>
Descriptive summary

Make filesets and Universal Viewer display user-contributed alternative text to improve accessibility of the public view.

Steps to reproduce the behavior in User Interface (UI)

  1.  Navigate to an image work that has an item/fileset with a thumbnail image.
  2.  Currently no alt text is available on the work's show page or in the UV

Acceptance Criteria/Expected Behavior

  *    Alt text available in item/fileset thumbnails
  *    Alt text available in Universal Viewer

Rationale (for feature request only)

Achieving WCAG 2.2 AA requires alt text for images beyond filename.

Related work

Bill Cox's explanation of OSU's implementation with attached patch files:

I went back and dusted off the commits for the alt text work and attached a few files:

  *   osul-alt-text.patch<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_user-2Dattachments_files_22784696_osul-2Dalt-2Dtext.patch&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=UnL802G5SdVV-32R3u-ZQCPy_3ItQxrnxxLobtXEX90&m=F0sAgN-UWc1AJzrRPMBVHsNdIsFUCeCo9y_nJN4_DTwQE7oQjNFzpU7NsdjZWFpu&s=wo3_WKvvSn0Ivnz0vZxPhIj-gXYW-kE1XetERy-ODj0&e=> haș the changes we made locally as an app built on hyrax
  *   hyrax-alt-text.patch<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_user-2Dattachments_files_22784697_hyrax-2Dalt-2Dtext.patch&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=UnL802G5SdVV-32R3u-ZQCPy_3ItQxrnxxLobtXEX90&m=F0sAgN-UWc1AJzrRPMBVHsNdIsFUCeCo9y_nJN4_DTwQE7oQjNFzpU7NsdjZWFpu&s=EtvXmw7Nd9uefRpuctcbsEbdlqb5aqSBB0euSH9aSjU&e=> has my untested attempt to translate that to hyrax itself
  *   uv-alt-text.patch<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_user-2Dattachments_files_22784695_uv-2Dalt-2Dtext.patch&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=UnL802G5SdVV-32R3u-ZQCPy_3ItQxrnxxLobtXEX90&m=F0sAgN-UWc1AJzrRPMBVHsNdIsFUCeCo9y_nJN4_DTwQE7oQjNFzpU7NsdjZWFpu&s=MJPeYhH_Kv3I6keg_WLpcv6vQV8za8bT_JtoH4YymMs&e=> has just our local changes to UV

On the front-end, we updated UV and the image view.
In UV, when creating the viewer <div>, we add an aria-label attribute with the alt-text. From the commit history I see that we also tried adding role=“img”, but we removed that after testing with Windows Narrator. To support this, we added the alt text to the IIIF::Presentation::ImageResource when creating the manifest.
In Rails, we overrode app/views/hyrax/file_sets/media_display/_image.html.erb to use file_set.alt_text and app/views/hyrax/file_sets/_form.html.erb to allow setting the alt text.

The rest of the plumbing was adding the property to ::FileSet and including it in to_solr, adding ::SolrDocument#alt_text to expose the indexed field, and creating our own ::FileSetPresenter to delegate the alt text to the solr document.

I saw online that there was some discussion about the predicate to use for the alt text. For our purposes, we decided to side-step that issue and created a dummy vocabulary:

VOCAB = ::RDF::Vocabulary.new("https://example.com/vocab/")
property :alt_text, predicate: VOCAB.alt_text, multiple: false


We also updated the configuration to include the alt text in the edit form and to use our presenter for filesets. The presenter configuration for filesets isn’t as centralized as it is for generic works, so we had to set it in multiple places and override Hyrax::Forms::WorkForm to allow configuring it.

Hyrax::Forms::FileSetEditForm.terms += %i[alt_text]

Hyrax::FileSetsController.show_presenter = ::FileSetPresenter
Hyrax::Forms::WorkForm.file_presenter_class = ::FileSetPresenter
Hyrax::MemberPresenterFactory.file_presenter_class = ::FileSetPresenter
Hyrax::PcdmMemberPresenterFactory.file_presenter_class = ::FileSetPresenter
Hyrax::SingleUseLinksViewerController.presenter_class = ::FileSetPresenter


We followed this pattern when added captions and transcripts.

The biggest issue is that this currently only sets the alt text once for the viewer. This handles a lot of our use cases but additional work is needed to handle multiple images in the same fileset.
